ext/lists.go

```diff
@@ -276,29 +450,122 @@ func flatten(list traits.Lister, depth int64) ([]ref.Val, error) {
 }
 
 func sortList(list traits.Lister) (ref.Val, error) {
+	return sortListByAssociatedKeys(list, list)
+}
+
+func sortListByAssociatedKeys(list traits.Lister, keys traits.Lister) (ref.Val, error) {
 	listLength := list.Size().(types.Int)
+	keysLength := keys.Size().(types.Int)
+	if listLength != keysLength {
+		return nil, fmt.Errorf(
+			"sortedByAssociatedKeys() expected a list of the same size as the associated keys list, but got %d and %d elements respectively",
+			listLength,
+			keysLength,
+		)
+	}
 	if listLength == 0 {
 		return list, nil
 	}
-	elem := list.Get(types.IntZero)
+	elem := keys.Get(types.IntZero)
 	if _, ok := elem.(traits.Comparer); !ok {
 		return nil, fmt.Errorf("list elements must be comparable")
 	}
 
+	sortedIndices := make([]ref.Val, 0, listLength)
+	for i := types.IntZero; i < listLength; i++ {
+		if keys.Get(i).Type() != elem.Type() {
+			return nil, fmt.Errorf("list elements must have the same type")
+		}
+		sortedIndices = append(sortedIndices, i)
+	}
+
+	sort.Slice(sortedIndices, func(i, j int) bool {
+		iKey := keys.Get(sortedIndices[i])
+		jKey := keys.Get(sortedIndices[j])
+		return iKey.(traits.Comparer).Compare(jKey) == types.IntNegOne
+	})
+
 	sorted := make([]ref.Val, 0, listLength)
+	for i := types.IntZero; i < listLength; i++ {
+		sorted = append(sorted, list.Get(sortedIndices[i]))
+	}
+	return types.DefaultTypeAdapter.NativeToValue(sorted), nil
+}
+
+// sortByMacro transforms an expression like:
+//
+//	mylistExpr.sortBy(e, -math.abs(e))
+//
+// into something equivalent to:
+//
+//	 cel.bind(
+//		  __sortBy_input__,
+//		  myListExpr,
+//		  __sortBy_input__.sortByAssociatedKeys(__sortBy_input__.map(e, -math.abs(e))
+//		)
+func sortByMacro(meh cel.MacroExprFactory, target ast.Expr, args []ast.Expr) (ast.Expr, *cel.Error) {
+	varIdent := meh.NewIdent("__sortBy_input__")
```

Comment:
Consider using `@` as a prefix to indicate the variable is not intended to be accessed.